### ISSUE TITLE
Sharding not working properly

### ISSUE THREAD

**Issue by jessesuen** [akuity/kargo#2791]
# Description

The following situation occurred because two controllers were reconciling the same Promotion

```yaml
status:
  phase: Succeeded
  message: 'failed to run step "argocd-update": Argo CD integration is disabled on
    this controller; cannot update Argo CD Application resources'
```

# Screenshots

<img width="653" alt="image" src="https://github.com/user-attachments/assets/cd3508f0-ee3d-4704-a3b6-2023468d5165">

# Steps to Reproduce

# Version

```
v1.0.0
```

# Logs

```
Paste any relevant application logs here.
```

**Comment by krancour** [comment]
@Marvin9 could correct me if I am wrong, but all the steps will show green if the Promotion appears to have been successful. The problem here seems _only_ to be that the Promotion was incorrectly counted as successful. I'm looking into where, _after_ the step failed (that step _does_ return a failure under these conditions) the Promo got deemed a success.

**Comment by Marvin9** [comment]
> but all the steps will show green if the Promotion appears to have been successful

The success state of any step depends on `currentStep` in promotion status. If UI sees that for any step n if n < `currentStep` then it is counted as success step. Let me know if that helps, or if logic needs to be changed

**Comment by krancour** [comment]
I don't think it's a UI bug. The YAML says the Promotion succeeded even though the step, for sure, returned an error.

**Comment by krancour** [comment]
Also, fwiw, I can force this error to happen by explicitly disabling Argo CD integration, but when I do that, both the step and the Promotion fail as expected.

**Comment by jessesuen** [comment]
This appears to have been caused by sharding logic not working. Updated title and description

**Comment by hiddeco** [comment]
It is confirmed to be solved by a manual build of #2817.
